virtualtime/datetime_tz/__init__.py

```diff
@@ -11,6 +11,9 @@
 
 class datetime_tz(base_datetime_tz.datetime_tz):
 
+  def __hash__(self):
+    return super(datetime_tz, self).__hash__()
+
```

Comment:
I'm a bit confused. If you don't define a method shouldn't it just call the method on the parent class?

Reply:
To answer my own question [If you define `__eq__` on a class `__hash__` is implicitly set to `None`](https://docs.python.org/3/reference/datamodel.html#object.__hash__). The rationale behind this is that if you redefined `__eq__` you most likely want to redefine `__hash__` to ensure that two equal values always hash the same. Hence python "undefines" `__hash__` so that the code won't accidentally use a "wrong" `__hash__` function from the parent class.

Reply:
If we dig into the datetime code we find the hash function:

``` python
    def __hash__(self):
        if self._hashcode == -1:
            if self.fold:
                t = self.replace(fold=0)
            else:
                t = self
            tzoff = t.utcoffset()
            if tzoff is None:
                self._hashcode = hash(t._getstate()[0])
            else:
                days = _ymd2ord(self.year, self.month, self.day)
                seconds = self.hour * 3600 + self.minute * 60 + self.second
                self._hashcode = hash(timedelta(days, seconds, self.microsecond) - tzoff)
        return self._hashcode
```
Which as far as I can tell hashes the `timedelta` since 1 January 0001 AD in Gregorian reckoning if the datetime object has a timezone and hashes serialized representation of the year if it doesn't have a timezone.

Further investigation in the datetime class shows that `__eq__` is defined such that naive datetimes and non-naive datetimes never compare equal.

However in our definition of `__eq__` naive datetimes and non-naive datetimes can compare equal (the naive datetime is localized before comparison). And so it's quite possible for `virtualtime.datetime_tz(...) == datetime(...)` but `hash(virtualtime_datetime_tz(...)) != hash(datetime(...))`, which violates the very invariant that the error was supposed to prevent.

The original `datetime_tz` code seems like it doesn't override `__eq__` to localize the second parameter (It throws an error in Python 2 and compares false in Python 3).

We are we localizing the second argument to `__eq__` in our virtualtime version of datetime_tz?

Reply:
I edited the previous comment to make it more clear. Read it on gitlab and not through your email.